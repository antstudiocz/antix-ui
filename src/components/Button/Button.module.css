/* Base button styles */
.button {
  padding: 14px 16px;
  border: none;
  border-radius: 12px;
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  transition: all 0.2s ease;
  letter-spacing: 0.2px;
  white-space: nowrap;
  text-align: center;
  outline: none;
  position: relative;
  overflow: hidden;
  font-family: var(--font-secondary);
  font-weight: 700;
  box-sizing: border-box;
  height: fit-content;
}

/* Sizes */
.xl {
  padding: 18px 28px;
  font-size: var(--font-button1-size);
  line-height: var(--font-button1-line-height);
}

.lg {
  padding: 16px 24px;
  font-size: var(--font-button1-size);
  line-height: var(--font-button1-line-height);
}

.md {
  padding: 12.5px 14px;
  font-size: var(--font-button2-size);
  line-height: var(--font-button2-line-height);
}

.sm {
  padding: 8px 14px;
  font-size: var(--font-button3-size);
  line-height: var(--font-button3-line-height);
}

/* Icon styles */
.withIcon {
  gap: 10px;
}

.withIconSm {
  gap: 6px;
}

.iconLeft {
  flex-direction: row;
}

.iconRight {
  flex-direction: row-reverse;
}

.icon {
  display: flex;
  align-items: center;
  justify-content: center;
}

/* ===== Variants ===== */
/* Solid variant */
.variantSolid {
  background-color: currentColor;
  color: var(--color-neutral-00);
  stroke: var(--color-neutral-00);

  & .buttonText,
  & .icon {
    color: var(--color-neutral-00);
    stroke: var(--color-neutral-00);
  }
}

/* Outlined variant */
.variantOutlined {
  background-color: transparent;
  outline: 1.67px solid currentColor;
  outline-offset: -1.67px;

  & .buttonText,
  & .icon {
    color: currentColor;
    stroke: currentColor;
  }

  &:hover:not(.disabled),
  &:active:not(.disabled) {
    & .buttonText,
    & .icon {
      color: var(--color-neutral-00);
      stroke: var(--color-neutral-00);
    }
  }
}

/* Text variant */
.variantText {
  background-color: transparent;
  padding: 2px;

  & .buttonText,
  & .icon {
    color: currentColor;
    stroke: currentColor;
    text-decoration: underline;
  }

  &:hover:not(.disabled),
  &:active:not(.disabled) {
    & .buttonText,
    & .icon {
      text-decoration: none;
    }
  }

  &.withIcon {
    & .buttonText,
    & .icon {
      text-decoration: none;
    }

    &:hover:not(.disabled),
    &:active:not(.disabled) {
      & .buttonText,
      & .icon {
        text-decoration: underline;
      }
    }
  }
}

/* ===== Colors ===== */
/* Conversion color (red) */
.colorConversion {
  color: var(--color-conversion-500);
  stroke: var(--color-conversion-500);

  &.variantSolid {
    &:hover:not(.disabled),
    &:active:not(.disabled) {
      background-color: var(--color-conversion-900);
    }
  }

  &.variantOutlined {
    &:hover:not(.disabled),
    &:active:not(.disabled) {
      background-color: var(--color-conversion-900);
      outline-color: var(--color-conversion-900);
    }
  }

  &.variantText {
    &:hover:not(.disabled),
    &:active:not(.disabled) {
      color: var(--color-conversion-900);
      stroke: var(--color-conversion-900);
    }
  }
}

/* Primary color (green) */
.colorPrimary {
  color: var(--color-primary-500);
  stroke: var(--color-primary-500);

  &.variantSolid {
    background: linear-gradient(
      to left,
      var(--color-primary-500),
      var(--color-highlight-300)
    );
    position: relative;
    z-index: 1;

    &::before {
      content: "";
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      background-color: var(--color-primary-700);
      opacity: 0;
      transition: opacity 0.2s ease;
      z-index: -1;
      border-radius: 12px;
    }

    &:hover:not(.disabled)::before,
    &:active:not(.disabled)::before {
      opacity: 1;
    }
  }

  &.variantOutlined {
    &:hover:not(.disabled),
    &:active:not(.disabled) {
      background-color: var(--color-primary-700);
      outline-color: var(--color-primary-700);
    }
  }

  &.variantText {
    &:hover:not(.disabled),
    &:active:not(.disabled) {
      color: var(--color-primary-700);
      stroke: var(--color-primary-700);
    }
  }
}

/* Secondary color (beige/green) */
.colorSecondary {
  color: var(--color-primary-500);
  stroke: var(--color-primary-500);

  &.variantSolid {
    background-color: var(--color-secondary-300);

    & .buttonText,
    & .icon {
      color: var(--color-primary-500);
      stroke: var(--color-primary-500);
    }

    &:hover:not(.disabled),
    &:active:not(.disabled) {
      background-color: var(--color-primary-700);

      & .buttonText,
      & .icon {
        color: var(--color-neutral-00);
        stroke: var(--color-neutral-00);
      }
    }
  }

  &.variantOutlined {
    background-color: var(--color-secondary-300);

    &:hover:not(.disabled),
    &:active:not(.disabled) {
      background-color: var(--color-primary-700);
      outline-color: var(--color-primary-700);
    }
  }

  &.variantText {
    &:hover:not(.disabled),
    &:active:not(.disabled) {
      color: var(--color-primary-700);
      stroke: var(--color-primary-700);
    }
  }
}

/* Other styles */
.buttonText {
  display: inline-block;
  white-space: nowrap;
  overflow: visible;
}

.fullWidth {
  width: 100%;
}

.disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

/* Special case for buttons with additional text */
.additionalText {
  font-size: var(--font-button3-size);
  line-height: var(--font-button3-line-height);
  font-weight: 700;
}

.minContent {
  min-width: fit-content;
  width: auto;
  flex-shrink: 0;
  max-width: 100%;
}
