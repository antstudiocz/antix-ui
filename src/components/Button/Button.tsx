import * as React from "react";
import { Slot } from "@radix-ui/react-slot";
import { cva, type VariantProps } from "class-variance-authority";

import { cn } from "@/lib/utils";

export type ButtonVariant = "solid" | "outlined" | "text";
export type ButtonColor = "conversion" | "primary" | "secondary";
export type ButtonSize = "xl" | "lg" | "md" | "sm";
export type ButtonIconPosition = "left" | "right";

// Base button styling with accurate values from the original .button
const buttonVariants = cva(
  "inline-flex items-center justify-center whitespace-nowrap rounded-xl font-secondary font-bold transition-all duration-200 text-center relative box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer",
  {
    variants: {
      variant: {
        solid: "", // Base styles for solid, specific colors are handled in combination with color
        outlined:
          "bg-transparent outline outline-[1.67px] outline-offset-[-1.67px] outline-current",
        text: "bg-transparent *:decoration-current",
      },
      size: {
        sm: "px-3.5 py-2",
        md: "px-3.5 py-3",
        lg: "px-6 py-3.5",
        xl: "px-7 py-4.5",
      },
      color: {
        conversion: "text-conversion-500",
        primary: "text-primary-500",
        secondary: "text-primary-500",
      },
      iconPosition: {
        left: "flex-row",
        right: "flex-row-reverse",
      },
      withIcon: {
        true: "", // Will be handled together with size
        false: "",
      },
    },
    // Default values
    defaultVariants: {
      variant: "solid",
      size: "md",
      color: "conversion",
      iconPosition: "left",
      withIcon: false,
    },
    // Complex variant combinations using conditional classes
    compoundVariants: [
      // Solid variant specific styles by color
      {
        variant: "solid",
        color: "conversion",
        className:
          "bg-conversion-500 *:text-neutral-00 hover:bg-conversion-900",
      },
      {
        variant: "solid",
        color: "primary",
        className:
          "relative bg-gradient-to-r to-primary-500 from-highlight-300 z-10 *:text-neutral-00 before:content-[''] before:absolute before:inset-0 before:bg-primary-700 before:opacity-0 before:transition-opacity before:duration-200 before:z-[-1] before:rounded-lg hover:before:opacity-100",
      },
      {
        variant: "solid",
        color: "secondary",
        className:
          "bg-secondary-300 *:text-primary-500 hover:bg-primary-700 hover:*:text-neutral-00",
      },
      // Outlined variant specific styles by color
      {
        variant: "outlined",
        color: "conversion",
        className:
          "hover:bg-conversion-900 hover:outline-conversion-900 hover:*:text-neutral-00",
      },
      {
        variant: "outlined",
        color: "primary",
        className:
          "hover:bg-primary-700 hover:outline-primary-700 hover:*:text-neutral-00",
      },
      {
        variant: "outlined",
        color: "secondary",
        className:
          "bg-secondary-300 hover:bg-primary-700 hover:outline-primary-700 hover:*:text-neutral-00",
      },
      // Text variant specific styles by color
      {
        variant: "text",
        color: "conversion",
        className: "hover:text-conversion-900",
      },
      {
        variant: "text",
        color: "primary",
        className: "hover:text-primary-700",
      },
      {
        variant: "text",
        color: "secondary",
        className: "hover:text-primary-700",
      },
      // Specific spacing for icons based on size
      {
        withIcon: true,
        size: ["sm", "md"],
        className: "gap-1.5",
      },
      {
        withIcon: true,
        size: ["lg", "xl"],
        className: "gap-2.5",
      },
      // Specific styles for text variant with icon
      {
        variant: "text",
        className: "*:no-underline hover:underline",
      },
      // Ensure text variant has no padding regardless of size
      {
        variant: "text",
        className: "!p-0",
      },
    ],
  }
);

export interface ButtonProps
  extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, "color">,
    Omit<VariantProps<typeof buttonVariants>, "className" | "withIcon"> {
  /** Allows the button to be rendered as a different element */
  asChild?: boolean;
  /** Icon element to be displayed alongside the button text */
  icon?: React.ReactNode;
  /** Additional text to be displayed below the main button text */
  additionalText?: string;
  /** Additional CSS classes to be applied to the button */
  className?: string;
}

/**
 * Button component with multiple variants, sizes, and colors
 *
 * @param variant - Visual style of the button ("solid", "outlined", "text")
 * @param size - Size of the button ("xl", "lg", "md", "sm")
 * @param color - Color scheme of the button ("conversion", "primary", "secondary")
 * @param iconPosition - Position of the icon relative to text ("left", "right")
 * @param asChild - Whether to render the button as a child component
 * @param icon - Icon element to display
 * @param additionalText - Secondary text to display below main text
 * @param children - Primary content of the button
 * @param disabled - Whether the button is disabled
 */
export const Button = ({
  className,
  variant,
  size = "md",
  color,
  iconPosition,
  asChild = false,
  icon,
  additionalText,
  children,
  disabled,
  ...props
}: ButtonProps) => {
  const Comp = asChild ? Slot : "button";

  /**
   * Important: Text size classes are applied directly through the cn() function instead of using variants in cva.
   *
   * Reason: When using variants in cva, conflicts may occur where other variants (e.g., color)
   * override text size classes. By extracting these classes from the variant system and applying them
   * directly in the cn() function, we ensure they are never overridden by other variants.
   */
  const getTextSizeClass = (size: ButtonSize) => {
    switch (size) {
      case "xl":
        return "text-button-xl";
      case "lg":
        return "text-button-lg";
      case "sm":
        return "text-button-sm";
      case "md":
      default:
        return "text-button-md";
    }
  };

  return (
    <Comp
      className={
        cn(
          buttonVariants({
            variant,
            size,
            color,
            iconPosition,
            withIcon: Boolean(icon),
          }),
          className
        ) +
        " " +
        getTextSizeClass(size || "md")
      }
      disabled={disabled}
      {...props}
    >
      {icon && (
        <span className="icon inline-flex items-center justify-center text-current">
          {icon}
        </span>
      )}
      {children && (
        <span className="inline-block whitespace-nowrap overflow-visible text-current">
          {children}
        </span>
      )}
      {additionalText && (
        <span className="text-button-sm font-bold text-current">
          {additionalText}
        </span>
      )}
    </Comp>
  );
};

export default Button;
