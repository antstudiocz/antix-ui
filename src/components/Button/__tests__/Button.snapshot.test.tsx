import React from "react";
import { render } from "@testing-library/react";
import { Button } from "../Button";

describe("Button Snapshots", () => {
  it("matches snapshot for solid variant", () => {
    const { container } = render(<Button variant="solid">Solid Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot for outlined variant", () => {
    const { container } = render(
      <Button variant="outlined">Outlined Button</Button>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot for text variant", () => {
    const { container } = render(<Button variant="text">Text Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot for different sizes", () => {
    const { container: smallContainer } = render(
      <Button size="sm">Small Button</Button>
    );
    expect(smallContainer).toMatchSnapshot();

    const { container: mediumContainer } = render(
      <Button size="md">Medium Button</Button>
    );
    expect(mediumContainer).toMatchSnapshot();

    const { container: largeContainer } = render(
      <Button size="lg">Large Button</Button>
    );
    expect(largeContainer).toMatchSnapshot();
  });

  it("matches snapshot for disabled state", () => {
    const { container } = render(<Button disabled>Disabled Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot for button with icon", () => {
    const { container } = render(
      <Button icon={<span>🔍</span>} iconPosition="left">
        With Icon
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot for button with right icon", () => {
    const { container } = render(
      <Button icon={<span>🔍</span>} iconPosition="right">
        With Right Icon
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot for button with additional text", () => {
    const { container } = render(
      <Button additionalText="Additional Info">Button with Text</Button>
    );
    expect(container).toMatchSnapshot();
  });
});
