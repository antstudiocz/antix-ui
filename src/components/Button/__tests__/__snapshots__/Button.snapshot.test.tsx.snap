// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshots matches snapshot for button with additional text 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Button with Text
    </span>
    <span
      class="text-button3 font-bold"
    >
      Additional Info
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for button with icon 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900 gap-1.5"
  >
    <span
      class="icon inline-flex items-center justify-center"
    >
      <span>
        üîç
      </span>
    </span>
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Button with Icon
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for button with right icon 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row-reverse text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900 gap-1.5"
  >
    <span
      class="icon inline-flex items-center justify-center"
    >
      <span>
        üîç
      </span>
    </span>
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Button with Right Icon
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for different sizes 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-2 flex-row text-button-sm bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Small Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for different sizes 2`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Medium Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for different sizes 3`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-6 py-3.5 flex-row text-button-lg bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Large Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for disabled state 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
    disabled=""
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Disabled Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for full width button 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Full Width Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for outlined variant 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current bg-transparent outline-[1.67px] outline-offset-[-1.67px] outline-current px-3.5 py-3 flex-row text-button-md hover:bg-conversion-900 hover:outline-conversion-900 hover:*:text-neutral-00 hover:*:stroke-neutral-00"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Outlined Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for solid variant 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current px-3.5 py-3 flex-row text-button-md bg-conversion-500 *:text-neutral-00 *:stroke-neutral-00 hover:bg-conversion-900"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Solid Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshots matches snapshot for text variant 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-lg font-secondary font-bold transition-all duration-200 text-center relative overflow-hidden box-border h-fit disabled:opacity-60 disabled:cursor-not-allowed cursor-pointer stroke-current bg-transparent p-[2px] *:decoration-current px-3.5 py-3 flex-row text-button-md hover:text-conversion-900 hover:stroke-conversion-900 *:underline hover:*:no-underline"
  >
    <span
      class="inline-block whitespace-nowrap overflow-visible"
    >
      Text Button
    </span>
  </button>
</div>
`;
